cognify/optimizer/trace/checkpoint.py

```diff
@@ -0,0 +1,409 @@
+            ):
+                continue
+            cancidates.append(log)
+        return cancidates
+
+    @property
+    def best_metrics_so_far(self):
+        return self.best_score, self.lowest_cost, self.fastest_exec_time
+    
+class LogManager:
+    _instance = None
+    
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super(LogManager, cls).__new__(cls)
+            cls._instance._init(*args, **kwargs)
+        return cls._instance
+    
+    def _init(self, log_dir):
+        # path to store the history of all real-evaluations
+        self.opt_trace_log_path = os.path.join(log_dir, "opt_trace.json")
+        # by-layer configuration logs
+        self.layer_stats: dict[str, LayerStat] = {}
+        # global best metrics
+        self._glob_best_score = CommonStats.base_quality
+        self._glob_lowest_cost = CommonStats.base_price
+        self._glob_fastest_exec_time = CommonStats.base_exec_time
+        self._glob_lock = threading.Lock()
+        self._opt_trace = []
+        # total cost of optimization
+        self.total_opt_cost = 0.0
+    
+    def register_layer(
+        self, 
+        layer_name: str,
+        layer_instance: str, 
+        opt_log_path: str, 
+        is_leaf: bool
+    ):
+        if layer_instance in self.layer_stats:
+            # raise ValueError(f"Layer {layer_instance} already registered at LogManager")
+            return
+        self.layer_stats[layer_instance] = LayerStat(layer_name, layer_instance, opt_log_path, is_leaf)
+    
+    def remove_layer(
+        self,
+        layer_instance: str
+    ):
+        del self.layer_stats[layer_instance]
+    
+    def add_trial(self, layer_instance: str, params) -> str:
+        """Add a proposed configuration to the layer log
+        """
+        return self.layer_stats[layer_instance].add_trial(params)
+
+    def num_trials(self, layer_instance: str, finished: bool = True):
+        return len(self.layer_stats[layer_instance].get_completed_logs()) if finished \
+                else len(self.layer_stats[layer_instance].opt_logs)
+    
+    def report_trial_result(self, layer_instance: str, id: str, result: EvaluationResult):
+        """Report the evaluation result of a configuration
+        """
+        if result is None or not result.complete:
+            return
+        self.layer_stats[layer_instance].report_result(id, result)
+        if self.layer_stats[layer_instance].is_leaf:
+            with self._glob_lock:
+                # update global
+                self._update_glob_best_score(result.reduced_score)
+                self._update_glob_lowest_cost(result.reduced_price)
+                self._update_glob_fastest_exec_time(result.reduced_exec_time)
+                self.total_opt_cost += result.total_eval_cost
+                
+                # save the evaluation trace
+                _result_cpy = copy.deepcopy(result)
+                _result_cpy.meta = {
```

Comment:
Set fields instead of dictionary keys

Reply:
Why not just define fields and then set those? Less error-prone

Reply:
Added a data structure for this.